### ISSUE TITLE
The next `Expr`

### ISSUE THREAD

**Issue by JacquesCarette** [JacquesCarette/Drasil#2482]
In some ways, this is more a discussion that @smiths and @balacij at least need to be involved in.

I was trying to remove the various `Discrete` types/spaces: they are not used, and not really implemented. And when we tried (last summer) to use them more, by generating 'enums' of various kinds for our languages, very serious difficulties arose. Basically, many of the languages require these to be declared, with a name, before they can be used. Which is reasonable enough. But Drasil doesn't really have any kind of notion of declaration, or of type definition. So I was going to remove that feature, since it's not compilable. It can be re-implemented later, when we have a proper design.

But I was foiled: `DiscreteS` is used. (`DiscreteD` appears in the code too, but it is unused). But I dug deeper: it is 'used' in a fundamental way https://github.com/JacquesCarette/Drasil/blob/b585428f1626a9a6aebd93bb7309869c7ed720d7/code/drasil-utils/Utils/Drasil/Fold.hs#L25-L32

But what is this really saying? It is sneaky: it is a hack for displaying somewhat meaningful stuff (`EnumeratedReal` and `EnumeratedStr`) in a more traditional math way, using the set-membership symbol.

The insight: we need at least 3 different languages of expressions:
1. mathematically-meaningful expressions, meant to be 'understood' by Drasil (reasoned about internally)
2. computationally-meaningful expressions,  meant to be translated to external programming languages
3. displayable expressions, meant for humans to puzzle out

There also needs to be a subset of 1 that can be translated to 2. All of 1 should be translatable to 3.  [Because the translatable ones might be a subset, that might force 1 to split into two languages.]

Our deep problem is that we currently have a single language. And this really makes things difficult, forcing us into hacks. It is holding Drasil back.

The difficult design is whether we have 4 totally separate languages and partial translators, or whether parts of the languages can be shared. A shared design is more elegant, but might be quite hard: we want the 'meaningful' expressions to be typed, while we can't for the displayable expressions. Furthermore, the type system for the computational parts may not necessarily be compatible - it would have to be coherent.  What I mean is that there would be a (bidirectional?) translation between those types but it might not be 'the identity'.

My instinct, even though it is computationally wasteful, is to go with 4 languages:
1. math, that can be reasoned about and computed with
2. math, that can be reasoned about (like statements in logic with forall, to express specification-level things)
3. computation-oriented
4. display-oriented

There should be a total translation from 1 to 2, 3 and 4. And a total translation from 2 to 4 as well. All of 3 should really be translatable to our supported PLs.  And all of 4 should be translatable to latex and HTML.

**Comment by balacij** [comment]
I think this is a very interesting issue! It's quite complex. I'm not sure if I'm fully understanding everything we want out of these 4 languages.

When you say
> There should be a total translation from 1 to 2, 3 and 4.

Are you also implying that we would never be hand-writing expressions in (2)? If so, what would (2) be used for other than converting into (4)? If not, I imagine we would be embedding expressions of (2) into (1), in which case, would we also be generating input/output validation/assertions in code using (2)?

**Comment by JacquesCarette** [comment]
I meant there should be an internal function that given a term in language (1), translates it to one in language (2) [and 2 more functions like it for (3) and (4)] and never fails.  i.e. language (1) should be a subset of the others.

Because there will be things in (2), (3) and (4) that are specific to each, those parts will still be hand-written. But writing something directly in (3) or (4) should feel bad, as it is admitting that it is (straight code, straight display information) that has "no meaning" as far as Drasil is concerned.

**Comment by balacij** [comment]
Hmm, I'm not sure I fully know what the state of this ticket is. We now have Expr, ModelExpr, and CodeExpr, but I'm not sure I understand which of 1/2/3/4 they fill. Expr could either be (1) or (3), CodeExpr could be (3) or a variant of it, and ModelExpr could be either (2) or (4)?

Regarding DiscreteS and DiscreteD, it looks like all of DiscreteS usage is removed, but [there is one usage of DiscreteD](https://jacquescarette.github.io/Drasil/examples/glassbr/SRS/srs/GlassBR_SRS.html#Table:ToS) that seems a bit odd. I don't imagine constraints should be appearing in the name of variables. Should we remove it? The actual restriction would need to wait for https://github.com/JacquesCarette/Drasil/issues/2655 (halted at the moment) to have it handled properly. If we were to remove it from the name, then I think we can remove both DiscreteS and DiscreteD.

**Comment by JacquesCarette** [comment]
I agree that resolving this fully is not needed for your thesis, so it can be removed from that project. And yes, the pseudo-design here doesn't fully line up with what we currently have. We can tackle that a little later.

**Comment by balacij** [comment]
Perfect, thank you!
